File: python/src/dealcode/_codec.py
"""The dealcode codec (SPEC.md format version 1)."""

from __future__ import annotations

import hashlib
from typing import Union

from ._alphabets import PRESETS, _ascii_lower, resolve
from ._ff1 import FF1

__all__ = ["Dealcode", "DealcodeError", "ConfigError", "RangeError", "InvalidCodeError"]

_COUNTER_BOUND = 2**63  # counters live in [0, min(radix**max_length, 2**63))
_CODESPACE_BOUND = 2**128  # radix**max_length must not exceed this
_TWEAK_PREFIX = "dealcode/v1/"
_KDF_PREFIX = b"dealcode/v1/kdf"


class DealcodeError(ValueError):
    """Base class for all dealcode errors."""


class ConfigError(DealcodeError):
    """Invalid codec configuration."""


class RangeError(DealcodeError):
    """Counter out of the encodable range."""


class InvalidCodeError(DealcodeError):
    """Code fails length, charset, or range validation."""


def _encode_clean_utf8(value: str, what: str) -> bytes:
    """UTF-8 encode, rejecting U+0000 and unpaired surrogates (SPEC.md §2.1)."""
    if "\x00" in value:
        raise ConfigError(f"{what} must not contain U+0000")
    try:
        return value.encode("utf-8")
    except UnicodeEncodeError as exc:
        raise ConfigError(f"{what} must be valid Unicode (no unpaired surrogates)") from exc


def _resolve_key(key: Union[bytes, bytearray, str]) -> bytes:
    """Key material handling (SPEC.md §2.1)."""
    if isinstance(key, str):
        if _ascii_lower(key) in PRESETS:
            raise ConfigError(
                f'string key "{key}" is a preset alphabet name '
                f"— did you swap the key and alphabet arguments?"
            )
        material = _encode_clean_utf8(key, "string key material")
        if not material:
            raise ConfigError("key must not be empty")
        return hashlib.sha256(_KDF_PREFIX + material).digest()
    if isinstance(key, (bytes, bytearray)):
        material = bytes(key)
        if not material:
            raise ConfigError("key must not be empty")
        if len(material) in (16, 24, 32):
            return material
        return hashlib.sha256(_KDF_PREFIX + material).digest()
    raise ConfigError("key must be bytes or a string")


def _default_max_length(radix: int, min_length: int) -> int:
    length = min_length
    cap = radix**min_length
    while cap * radix < _COUNTER_BOUND:  # largest L with radix**L <= 2**63 - 1
        cap *= radix
        length += 1
    return length


class Dealcode:
    """Bijective counter <-> code mapping. See SPEC.md.

    >>> codec = Dealcode(key="example-key")
    >>> codec.decode(codec.encode(42)) == 42
    True

    Instances are immutable, thread-safe, and cheap to keep around; create one
    per code namespace at startup and reuse it.
    """

    __slots__ = (
        "_alphabet",
        "_radix",
        "_min_length",
        "_max_length",
        "_domain",
        "_tweak",
        "_ff1",
        "_index",
        "_powers",
        "_capacity",
    )

    def __init__(
        self,
        key: Union[bytes, bytearray, str],
        alphabet: str = "hex",
        *,
        min_length: int = 6,
        max_length: Union[int, None] = None,
        domain: str = "",
    ) -> None:
        aes_key = _resolve_key(key)
        try:
            self._alphabet = resolve(alphabet)
        except ValueError as exc:
            raise ConfigError(str(exc)) from exc
        radix = len(self._alphabet.chars)

        # Bound the lengths BEFORE computing any power (SPEC.md §2): 128 is
        # the structural maximum (radix >= 2, radix**max_length <= 2^128).
        if not isinstance(min_length, int) or isinstance(min_length, bool) or not (
            2 <= min_length <= 128
        ):
            raise ConfigError("min_length must be an integer in [2, 128]")
        if radix**min_length < 100:
            raise ConfigError(
                "radix**min_length must be at least 100 (FF1 minimum domain)"
            )
        if max_length is None:
            max_length = _default_max_length(radix, min_length)
        if not isinstance(max_length, int) or isinstance(max_length, bool) or not (
            min_length <= max_length <= 128
        ):
            raise ConfigError("max_length must be an integer in [min_length, 128]")
        if radix**max_length > _CODESPACE_BOUND:
            raise ConfigError("radix**max_length must not exceed 2^128")
        if not isinstance(domain, str):
            raise ConfigError("domain must be a string")
        if len(_encode_clean_utf8(domain, "domain")) > 255:
            raise ConfigError("domain must be at most 255 UTF-8 bytes")

        self._radix = radix
        self._min_length = min_length
        self._max_length = max_length
        self._domain = domain
        self._tweak = (_TWEAK_PREFIX + domain).encode("utf-8")
        self._ff1 = FF1(aes_key, radix)
        self._index = {c: i for i, c in enumerate(self._alphabet.chars)}
        # powers[d] = radix**d for d in [0, max_length]
        self._powers = [radix**d for d in range(max_length + 1)]
        self._capacity = min(self._powers[max_length], _COUNTER_BOUND)

    # -- introspection -------------------------------------------------------

    @property
    def alphabet(self) -> str:
        return self._alphabet.chars

    @property
    def radix(self) -> int:
        return self._radix

    @property
    def min_length(self) -> int:
        return self._min_length

    @property
    def max_length(self) -> int:
        return self._max_length

    @property
    def domain(self) -> str:
        return self._domain

    @property
    def capacity(self) -> int:
        """Number of encodable counters: min(radix ** max_length, 2**63)."""
        return self._capacity

    def __repr__(self) -> str:  # keys never appear in repr/logs
        name = self._alphabet.name or f"custom({self._radix})"
        return (
            f"Dealcode(alphabet={name!r}, min_length={self._min_length}, "
            f"max_length={self._max_length}, domain={self._domain!r})"
        )

    # -- public API ------------------------------------------------------------

    def encode(self, n: int) -> str:
        """Map counter ``n`` to its code. O(1); raises RangeError if out of range."""
        if not isinstance(n, int) or isinstance(n, bool):
            raise RangeError(f"counter must be an integer, got {type(n).__name__}")
        if n < 0 or n >= self._capacity:
            raise RangeError(f"counter {n} out of range [0, {self._capacity})")
        d, base = self._stage_of(n)
        numerals = self._to_numerals(n - base, d)
        cipher = self._ff1.encrypt(self._tweak, numerals)
        chars = self._alphabet.chars
        return "".join(chars[x] for x in cipher)

    def decode(self, code: str) -> int:
        """Map a code back to its counter. Raises InvalidCodeError if not issuable."""
        if not isinstance(code, str):
            raise InvalidCodeError(f"code must be a string, got {type(code).__name__}")
        # Length gate before normalization: normalization is length-preserving,
        # so this is behaviour-identical, and it keeps rejection of oversized
        # garbage O(1) instead of normalizing megabytes first (SPEC §7).
        d = len(code)
        if d < self._min_length or d > self._max_length:
            raise InvalidCodeError(
                f"code length {d} outside [{self._min_length}, {self._max_length}]"
            )
        normalized = self._alphabet.normalize(code)
        index = self._index
        try:
            numerals = [index[c] for c in normalized]
        except KeyError as exc:
            raise InvalidCodeError(
                f"character {exc.args[0]!r} not in alphabet"
            ) from exc
        plain = self._ff1.decrypt(self._tweak, numerals)
        v = 0
        for x in plain:
            v = v * self._radix + x
        base = 0 if d == self._min_length else self._powers[d - 1]
        if d > self._min_length and v >= self._powers[d] - base:
            raise InvalidCodeError("code was not issued by this codec")
        n = base + v
        if n >= _COUNTER_BOUND:
            raise InvalidCodeError("code was not issued by this codec")
        return n

    # -- helpers ---------------------------------------------------------------

    def _stage_of(self, n: int) -> "tuple[int, int]":
        powers, m = self._powers, self._min_length
        if n < powers[m]:
            return m, 0
        d = m + 1
        while n >= powers[d]:
            d += 1
        return d, powers[d - 1]

    def _to_numerals(self, value: int, m: int) -> "list[int]":
        out = [0] * m
        r = self._radix
        for i in range(m - 1, -1, -1):
            value, out[i] = divmod(value, r)
        return out

File: js/src/codec.ts
/** The dealcode codec (SPEC.md format version 1). */

import { createHash } from "node:crypto";

import {
  matchesPresetName,
  resolveAlphabet,
  type ResolvedAlphabet,
} from "./alphabets.js";
import { ConfigError, CounterRangeError, InvalidCodeError } from "./errors.js";
import { FF1 } from "./ff1.js";

/** Counters live in `[0, min(radix^maxLength, 2^63))`. */
const COUNTER_BOUND = 2n ** 63n;
/** `radix^maxLength` must not exceed this. */
const CODESPACE_BOUND = 2n ** 128n;
const TWEAK_PREFIX = "dealcode/v1/";
const KDF_PREFIX = Buffer.from("dealcode/v1/kdf", "ascii");

/** Options for {@link Dealcode}. */
export interface DealcodeOptions {
  /**
   * Key material (required). A `Uint8Array` (or `Buffer`) of exactly 16, 24,
   * or 32 bytes is used directly as the AES key; any other non-empty bytes,
   * or any non-empty string (its UTF-8 bytes), are deterministically derived
   * into an AES-256 key via `SHA-256("dealcode/v1/kdf" ‖ material)`.
   * A hex-looking string is *not* auto-decoded. Generate a good key with
   * `openssl rand -hex 32`.
   */
  key: string | Uint8Array;
  /**
   * A preset name (`"dec"`, `"hex"`, `"base32"`, `"crockford"`, `"base36"`,
   * `"base58"`, `"base62"`, `"base64url"`) or a custom alphabet string of
   * 2–94 distinct printable ASCII characters. Default: `"hex"`.
   */
  alphabet?: string;
  /**
   * Codes start at this length (default 6). Must satisfy `minLength >= 2`
   * and `radix^minLength >= 100`.
   */
  minLength?: number;
  /**
   * Codes grow one character at a time up to this length. Default: the
   * largest `L` with `radix^L <= 2^63 − 1` (hex → 15, dec → 18, base62 → 10,
   * …). Must satisfy `minLength <= maxLength` and `radix^maxLength <= 2^128`.
   */
  maxLength?: number;
  /**
   * Namespace label bound into the FF1 tweak (default `""`). Two codecs with
   * the same key but different domains produce unrelated codes. At most 255
   * UTF-8 bytes.
   */
  domain?: string;
}

/**
 * Rejects U+0000 and unpaired surrogates in string inputs (SPEC.md §2.1) —
 * silently re-encoding them (Buffer would emit U+FFFD) makes the "same"
 * input produce different permutations across languages.
 */
function assertCleanUnicode(value: string, what: string): void {
  if (value.includes("\u0000")) {
    throw new ConfigError(`${what} must not contain U+0000`);
  }
  // String.prototype.isWellFormed needs lib >= ES2024; feature-detect via a
  // cast so the ES2022 build target still compiles, with a regex fallback.
  const isWellFormed = (value as { isWellFormed?: () => boolean }).isWellFormed;
  const wellFormed =
    typeof isWellFormed === "function"
      ? isWellFormed.call(value)
      : !/[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:^|[^\uD800-\uDBFF])[\uDC00-\uDFFF]/.test(value);
  if (!wellFormed) {
    throw new ConfigError(`${what} must be valid Unicode (no unpaired surrogates)`);
  }
}

/** Key material handling (SPEC.md §2.1). */
function resolveKey(key: string | Uint8Array): Buffer {
  let material: Buffer;
  let direct: boolean;
  if (typeof key === "string") {
    if (matchesPresetName(key)) {
      throw new ConfigError(
        `string key "${key}" is a preset alphabet name ` +
          `— did you swap the key and alphabet arguments?`,
      );
    }
    assertCleanUnicode(key, "string key material");
    material = Buffer.from(key, "utf8");
    direct = false; // strings are always derived, never auto-decoded
  } else if (key instanceof Uint8Array) {
    material = Buffer.from(key);
    direct = material.length === 16 || material.length === 24 || material.length === 32;
  } else {
    throw new ConfigError("key must be a string or a Uint8Array");
  }
  if (material.length === 0) {
    throw new ConfigError("key must not be empty");
  }
  if (direct) return material;
  return createHash("sha256").update(KDF_PREFIX).update(material).digest();
}

/** Largest L >= minLength with radix^L <= 2^63 − 1 (SPEC.md §2). */
function defaultMaxLength(radix: bigint, minLength: number): number {
  let length = minLength;
  let cap = radix ** BigInt(minLength);
  while (cap * radix < COUNTER_BOUND) {
    cap *= radix;
    length += 1;
  }
  return length;
}

/**
 * Bijective counter ↔ code mapping (dealcode format v1, see SPEC.md).
 *
 * Maps a non-negative integer counter to a short, random-looking,
 * fixed-alphabet code and back, using FF1 format-preserving encryption
 * (NIST SP 800-38G). The mapping is a keyed permutation: distinct counters
 * always produce distinct codes.
 *
 * ```ts
 * import { Dealcode } from "dealcode";
 *
 * const codec = new Dealcode({ key: process.env.DEALCODE_KEY! });
 * const code = codec.encode(42);   // e.g. "4b71b7"
 * codec.decode(code);              // 42n
 * ```
 *
 * Instances are immutable and safe for concurrent use; create one per code
 * namespace at startup and reuse it. The whole configuration (key, alphabet,
 * lengths, domain) must never change once codes have been issued.
 */
export class Dealcode {
  /** The alphabet characters, in numeral order. */
  readonly alphabet: string;
  /** Number of characters in the alphabet. */
  readonly radix: number;
  /** Length of the shortest codes. */
  readonly minLength: number;
  /** Length of the longest codes. */
  readonly maxLength: number;
  /** Namespace label bound into the FF1 tweak. */
  readonly domain: string;
  /** Number of encodable counters: `min(radix^maxLength, 2^63)`. */
  readonly capacity: bigint;

  readonly #alpha: ResolvedAlphabet;
  readonly #tweak: Buffer;
  readonly #ff1: FF1;
  readonly #charIndex: Map<string, number>;
  /** powers[d] = radix^d for d in [0, maxLength]. */
  readonly #powers: readonly bigint[];
  readonly #radixBig: bigint;

  constructor(options: DealcodeOptions) {
    if (options === null || typeof options !== "object") {
      throw new ConfigError("Dealcode requires an options object with a key");
    }
    const { key, alphabet = "hex", minLength = 6, maxLength, domain = "" } = options;
    if (key === undefined) {
      throw new ConfigError("key is required");
    }
    const aesKey = resolveKey(key);
    let alpha: ResolvedAlphabet;
    try {
      alpha = resolveAlphabet(alphabet);
    } catch (err) {
      throw new ConfigError((err as Error).message);
    }
    const radix = alpha.chars.length;
    const radixBig = BigInt(radix);

    // Bound lengths BEFORE computing any BigInt power (SPEC §2): 128 is the
    // structural maximum, and an unguarded exponent on absurd lengths would
    // either burn CPU or leak a raw V8 "Maximum BigInt size exceeded" error.
    if (!Number.isInteger(minLength) || minLength < 2 || minLength > 128) {
      throw new ConfigError("minLength must be an integer in [2, 128]");
    }
    if (radixBig ** BigInt(minLength) < 100n) {
      throw new ConfigError(
        `radix^minLength must be at least 100 (FF1 minimum domain); ` +
          `got ${radix}^${minLength}`,
      );
    }
    const max = maxLength ?? defaultMaxLength(radixBig, minLength);
    if (!Number.isInteger(max) || max < minLength || max > 128) {
      throw new ConfigError("maxLength must be an integer in [minLength, 128]");
    }
    if (radixBig ** BigInt(max) > CODESPACE_BOUND) {
      throw new ConfigError(
        `radix^maxLength must not exceed 2^128; got ${radix}^${max}`,
      );
    }
    if (typeof domain !== "string") {
      throw new ConfigError("domain must be a string");
    }
    assertCleanUnicode(domain, "domain");
    const tweak = Buffer.from(TWEAK_PREFIX + domain, "utf8");
    if (tweak.length - TWEAK_PREFIX.length > 255) {
      throw new ConfigError("domain must be at most 255 UTF-8 bytes");
    }

    this.alphabet = alpha.chars;
    this.radix = radix;
    this.minLength = minLength;
    this.maxLength = max;
    this.domain = domain;
    this.#alpha = alpha;
    this.#tweak = tweak;
    this.#ff1 = new FF1(aesKey, radix);
    this.#charIndex = new Map();
    for (let i = 0; i < alpha.chars.length; i++) {
      this.#charIndex.set(alpha.chars[i]!, i);
    }
    const powers: bigint[] = [1n];
    for (let d = 1; d <= max; d++) powers.push(powers[d - 1]! * radixBig);
    this.#powers = powers;
    this.#radixBig = radixBig;
    const codespace = powers[max]!;
    this.capacity = codespace < COUNTER_BOUND ? codespace : COUNTER_BOUND;
    Object.freeze(this);
  }

  /**
   * Map counter `n` to its code. O(1) in the counter value.
   *
   * @param n - A non-negative integer in `[0, capacity)`. A `number` must be
   *   a safe integer (use `bigint` beyond `Number.MAX_SAFE_INTEGER`).
   * @returns The code string; its length is between `minLength` and
   *   `maxLength`, growing only when the previous length is exhausted.
   * @throws {@link CounterRangeError} if `n` is not an integer or is out of
   *   range.
   */
  encode(n: number | bigint): string {
    let counter: bigint;
    if (typeof n === "bigint") {
      counter = n;
    } else if (typeof n === "number" && Number.isSafeInteger(n)) {
      counter = BigInt(n);
    } else if (typeof n === "number") {
      throw new CounterRangeError(
        `counter must be a safe integer; got ${n} (use a bigint for large counters)`,
      );
    } else {
      throw new CounterRangeError(`counter must be a number or bigint, got ${typeof n}`);
    }
    if (counter < 0n || counter >= this.capacity) {
      throw new CounterRangeError(`counter ${counter} out of range [0, ${this.capacity})`);
    }
    const [d, base] = this.#stageOf(counter);
    const numerals = this.#toNumerals(counter - base, d);
    const cipher = this.#ff1.encrypt(this.#tweak, numerals);
    let code = "";
    for (const x of cipher) code += this.alphabet[x];
    return code;
  }

  /**
   * Map a code back to its counter.
   *
   * Preset-alphabet normalization (e.g. case folding for `hex`, `O→0` for
   * `crockford`) is applied to the input first; custom alphabets must match
   * exactly.
   *
   * @returns The counter as a `bigint` (counters can exceed
   *   `Number.MAX_SAFE_INTEGER`; see {@link decodeNumber}).
   * @throws {@link InvalidCodeError} if the code has the wrong length,
   *   contains characters outside the alphabet, or was never issued by this
   *   codec. Decode success only proves consistency with the key — the
   *   application still decides whether counter `n` actually exists.
   */
  decode(code: string): bigint {
    if (typeof code !== "string") {
      throw new InvalidCodeError(`code must be a string, got ${typeof code}`);
    }
    // Length gate before normalization: normalization is length-preserving,
    // so this is behaviour-identical, and it keeps rejection of oversized
    // garbage O(1) instead of normalizing megabytes first (SPEC §7). The
    // gate counts code points when that is cheap so the message matches the
    // other implementations; valid codes are ASCII, where the counts agree.
    const gate =
      code.length <= 4 * this.maxLength ? [...code].length : code.length;
    if (gate < this.minLength || gate > this.maxLength) {
      throw new InvalidCodeError(
        `code length ${gate} outside [${this.minLength}, ${this.maxLength}]`,
      );
    }
    const normalized = this.#alpha.normalize(code);
    const d = normalized.length;
    const numerals = new Array<number>(d);
    for (let i = 0; i < d; i++) {
      const numeral = this.#charIndex.get(normalized[i]!);
      if (numeral === undefined) {
        throw new InvalidCodeError(
          `character ${JSON.stringify(normalized[i])} not in alphabet`,
        );
      }
      numerals[i] = numeral;
    }
    const plain = this.#ff1.decrypt(this.#tweak, numerals);
    let v = 0n;
    for (const x of plain) v = v * this.#radixBig + BigInt(x);
    const base = d === this.minLength ? 0n : this.#powers[d - 1]!;
    if (d > this.minLength && v >= this.#powers[d]! - base) {
      throw new InvalidCodeError("code was not issued by this codec");
    }
    const n = base + v;
    if (n >= COUNTER_BOUND) {
      throw new InvalidCodeError("code was not issued by this codec");
    }
    return n;
  }

  /**
   * Like {@link decode}, but returns a `number` for convenience.
   *
   * @throws {@link CounterRangeError} if the decoded counter exceeds
   *   `Number.MAX_SAFE_INTEGER` (use {@link decode} in that case).
   * @throws {@link InvalidCodeError} if the code is invalid (see
   *   {@link decode}).
   */
  decodeNumber(code: string): number {
    const n = this.decode(code);
    if (n > BigInt(Number.MAX_SAFE_INTEGER)) {
      throw new CounterRangeError(
        `decoded counter ${n} exceeds Number.MAX_SAFE_INTEGER; use decode() instead`,
      );
    }
    return Number(n);
  }

  // -- helpers ----------------------------------------------------------------

  /** Stage of counter n: [code length d, stage base]. */
  #stageOf(n: bigint): [number, bigint] {
    const powers = this.#powers;
    const m = this.minLength;
    if (n < powers[m]!) return [m, 0n];
    let d = m + 1;
    while (n >= powers[d]!) d += 1;
    return [d, powers[d - 1]!];
  }

  /** STR: represent value as exactly m base-radix numerals, big-endian. */
  #toNumerals(value: bigint, m: number): number[] {
    const out = new Array<number>(m);
    const r = this.#radixBig;
    let v = value;
    for (let i = m - 1; i >= 0; i--) {
      out[i] = Number(v % r);
      v /= r;
    }
    return out;
  }
}
